src/org/etools/j1939_84/controllers/part12/Part12Step12Controller.java

```diff
@@ -0,0 +1,248 @@
+                                                             .getEither()
+                                                             .stream()
+                                                             .peek(p -> {
+                                                                 String moduleName = Lookup.getAddressName(module.getSourceAddress());
+                                                                 if (p.right.isPresent()) {
+                                                                     // 12.12.a. Fail if NACK received for DM7 PG from
+                                                                     // OBD ECU
+                                                                     addFailure("6.12.12.a - NACK received for DM7 PG from OBD ECU "
+                                                                             + moduleName);
+                                                                 }
+                                                                 if (p.left.isPresent()) {
+                                                                     DM58RationalityFaultSpData packet = p.resolve();
+                                                                     // c. Fail, if expected unused bytes in DM58 are
+                                                                     // not padded with FFh
+                                                                     if (!areUnusedBytesPaddedWithFFh(packet)) {
+                                                                         addFailure("6.12.12.c - Unused bytes in DM58 are not padded with FFh in the response from "
+                                                                                 + moduleName);
+                                                                     }
+                                                                     // d. Fail, if data returned is greater than FBh
+                                                                     // (for 1 byte SP), FBFFh (for 2 byte SP), or
+                                                                     // FBFFFFFFh (for 4 byte SP)
+                                                                     if (isGreaterThanFb(packet)) {
+                                                                         addFailure("6.12.12.d - Data returned is greater than 0xFB... threshold");
+                                                                     }
+                                                                 }
+                                                             })
+                                                             .collect(Collectors.toList());
+
+                      // b. Fail, if DM58 not received (after allowed retries)
+                      if (packets.isEmpty()) {
+                          addFailure("6.12.12.d - DM58 not received (after allowed retries)");
```

Comment:
Since this is per module, the message should indicate which module didn’t respond